### ISSUE TITLE
rootpy fails creating dictionaries 

### ISSUE THREAD

**Issue by mverzett** [rootpy/rootpy#653]
Dear rootpyers,

I have recently moved to the HEAD of rootpy to have all the nice roofit/roostats features, but I found the following issue:

```
>>> from rootpy.stats.collection import ArgSet
WARNING:rootpy.utils.lock] 16921@cmslpc26.fnal.gov broke lock on /uscms/home/verzetti/.cache/rootpy/x86_64-53418/dicts/lock that is 63 seconds old
INFO:rootpy.stl] generating dictionary for deque<RooAbsArg*> ...
INFO:ROOT.TUnixSystem.ACLiC] creating shared library /uscms/home/verzetti/.cache/rootpy/x86_64-53418/dicts/d834ad667ff712a7.so
In file included from /tmp/rootcint_X1MaIL.h:3:0,
                 from /tmp/rJNUOW_cint.cxx:1:
/uscms/home/verzetti/.cache/rootpy/x86_64-53418/dicts/d834ad667ff712a7.C:3:24: fatal error: RooRealVar.h: No such file or directory
 #include "RooRealVar.h"
                        ^
compilation terminated.
Error: external preprocessing failed. :0:
!!!Removing /uscms/home/verzetti/.cache/rootpy/x86_64-53418/dicts/d834ad667ff712a7_ACLiC_dict.cxx /uscms/home/verzetti/.cache/rootpy/x86_64-53418/dicts/d834ad667ff712a7_ACLiC_dict.h !!!
Error: /cvmfs/cms.cern.ch/slc6_amd64_gcc481/lcg/root/5.34.18-cms4/bin/rootcint: error loading headers...
ERROR:ROOT.ACLiC] Dictionary generation failed!
INFO:ROOT.ACLiC] Invoking compiler to check macro's validity
/uscms/home/verzetti/.cache/rootpy/x86_64-53418/dicts/d834ad667ff712a7.C:3:24: fatal error: RooRealVar.h: No such file or directory
 #include "RooRealVar.h"
                        ^
compilation terminated.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/uscms_data/d3/verzetti/CMSSW_7_1_5/src/URAnalysis/external/src/rootpy/rootpy/stats/__init__.py", line 29, in <module>
    headers='<stack>;<deque>;RooRealVar.h')
  File "/uscms_data/d3/verzetti/CMSSW_7_1_5/src/URAnalysis/external/src/rootpy/rootpy/stl.py", line 426, in __call__
    cpptype.ensure_built(headers=headers)
  File "/uscms_data/d3/verzetti/CMSSW_7_1_5/src/URAnalysis/external/src/rootpy/rootpy/stl.py", line 204, in ensure_built
    child.ensure_built(headers=headers)
  File "/uscms_data/d3/verzetti/CMSSW_7_1_5/src/URAnalysis/external/src/rootpy/rootpy/stl.py", line 208, in ensure_built
    has_iterators=self.name in HAS_ITERATORS)
  File "/uscms_data/d3/verzetti/CMSSW_7_1_5/src/URAnalysis/external/src/rootpy/rootpy/stl.py", line 394, in generate
    sourcepath, 'k-', libname, DICTS_PATH) != 1):
  File "/uscms_data/d3/verzetti/CMSSW_7_1_5/src/URAnalysis/external/src/rootpy/rootpy/stl.py", line 394, in generate
    sourcepath, 'k-', libname, DICTS_PATH) != 1):
  File "/uscms_data/d3/verzetti/CMSSW_7_1_5/src/URAnalysis/external/src/rootpy/rootpy/logger/magic.py", line 247, in intercept_next_line
    raise exception
rootpy.ROOTError: level=3000, loc='ACLiC', msg='Dictionary generation failed!'
```

The error originates from [this line](https://github.com/urcms/rootpy/blob/master/rootpy/stats/__init__.py#L28). 

What is it needed for?
At the moment I have commented it out in my personal area and I do not have any trouble whatsoever.
 
I believe it is due to my ROOT settings, as otherwise it would fail your build testing, nevertheless it would be nice to overcome or at least protect against such failure and keep running without the dictionaries.

My Settings:
Python 2.7.6 
ROOT 5.34/18 -- embedded in CMSSW (official CMS software)
Any further information needed?

Thanks

**Comment by ndawe** [comment]
Thanks for reporting this. It looks like RooRealVar.h is not found. Does the ROOT build have RooStats enabled?

Testing myself with this ROOT release I don't see this problem:

```python
>>> from rootpy.stats.collection import ArgSet

RooFit v3.60 -- Developed by Wouter Verkerke and David Kirkby 
                Copyright (C) 2000-2013 NIKHEF, University of California & Stanford University
                All rights reserved, please read http://roofit.sourceforge.net/license.txt

INFO:rootpy.stl] generating dictionary for deque<RooAbsArg*> ...
INFO:ROOT.TUnixSystem.ACLiC] creating shared library /home/edawe/.cache/rootpy/x86_64-53418/dicts/d834ad667ff712a7.so
INFO:rootpy.stl] generating dictionary for stack<RooAbsArg*, deque<RooAbsArg*> > ...
INFO:ROOT.TUnixSystem.ACLiC] creating shared library /home/edawe/.cache/rootpy/x86_64-53418/dicts/deb57d21332feb79.so
>>> 
```

**Comment by kreczko** [comment]
Just tried
https://github.com/BristolTopGroup/DailyPythonScripts/blob/master/examples/roofit/roofit_advanced.py
on vanilla and it works.

However, a simple C++ test program:
```c++
#include "RooRealVar.h"

int main(){
}
```
fails.
```bash
scram setup roostats
# or
scram setup roofit
```
does not help. This header file is not included in ```/cvmfs/cms.cern.ch/slc6_amd64_gcc481/lcg/root/5.34.18-cms4/include/```

**Comment by kreczko** [comment]
On a side note:
The C++ example does not compile without any modification in ```CMSSW_7_5_0_pre5```(ROOT6) but at least ```RooRealVar.h``` is available in the include directory that ```root-config --incdir``` points to.

**Comment by mverzett** [comment]
Back to my previous question:
what are these dictionaries needed for?

**Comment by mverzett** [comment]
I have made a PR that solves the problem (#654). I added the option to permanently add include paths when generating dictionaries: [rootpy-add-include-path](https://github.com/mverzett/rootpy/blob/master/scripts/rootpy-add-include-path)

The script writes the paths in a text file that is read by stl.py which takes care of adding them in the ROOT interpreter.
For the record: CMS uses a ROOT version without RooFit and then adds a custom RooFit. The path is stored in $ROOFITSYS

**Comment by ndawe** [comment]
Thanks for the PR. Reviewing it. These dictionaries are required for internals of RooAbsArg and I think are generated/loaded here to avoid seeing warnings about missing dictionaries when retrieving Roo objects from a file. I don't think they are strictly required by any functionality in rootpy, however.

**Comment by ndawe** [comment]
Closed by #654
